File: frontend/src/types/setting.ts
export interface UserSetting {
  id: number
  user_id: number
  llm_provider: string | null
  llm_api_key: string | null
  llm_api_base_url: string | null
  llm_model_name: string | null
  ollama_base_url: string | null
  ollama_model: string | null
  tool_llm_provider: string | null
  tool_llm_api_key: string | null
  tool_llm_api_base_url: string | null
  tool_llm_model_name: string | null
  tool_ollama_base_url: string | null
  tool_ollama_model: string | null
  intent_llm_provider: string | null
  intent_llm_api_key: string | null
  intent_llm_api_base_url: string | null
  intent_llm_model_name: string | null
  emotion_llm_provider: string | null
  emotion_llm_api_key: string | null
  emotion_llm_api_base_url: string | null
  emotion_llm_model_name: string | null
  format_llm_provider: string | null
  format_llm_api_key: string | null
  format_llm_api_base_url: string | null
  format_llm_model_name: string | null
  show_tool_result: boolean | null
  show_collaboration: boolean | null
  token_monthly_budget: number | null
  created_at: string
  updated_at: string
}

export interface SettingUpdate {
  llm_provider?: string
  llm_api_key?: string
  llm_api_base_url?: string
  llm_model_name?: string
  ollama_base_url?: string
  ollama_model?: string
  tool_llm_provider?: string
  tool_llm_api_key?: string | null
  tool_llm_api_base_url?: string | null
  tool_llm_model_name?: string | null
  tool_ollama_base_url?: string | null
  tool_ollama_model?: string | null
  intent_llm_provider?: string | null
  intent_llm_api_key?: string | null
  intent_llm_api_base_url?: string | null
  intent_llm_model_name?: string | null
  emotion_llm_provider?: string | null
  emotion_llm_api_key?: string | null
  emotion_llm_api_base_url?: string | null
  emotion_llm_model_name?: string | null
  format_llm_provider?: string | null
  format_llm_api_key?: string | null
  format_llm_api_base_url?: string | null
  format_llm_model_name?: string | null
  show_tool_result?: boolean | null
  show_collaboration?: boolean | null
  token_monthly_budget?: number | null
}

export interface TokenUsageSummary {
  today_tokens: number
  month_tokens: number
  total_tokens: number
  monthly_budget: number | null
}

File: backend/app/schemas/setting.py
from datetime import datetime

from pydantic import BaseModel, field_validator


class SettingUpdate(BaseModel):
    llm_provider: str | None = None
    llm_api_key: str | None = None
    llm_api_base_url: str | None = None
    llm_model_name: str | None = None
    ollama_base_url: str | None = None
    ollama_model: str | None = None
    # Tool Generation AI config
    tool_llm_provider: str | None = None
    tool_llm_api_key: str | None = None
    tool_llm_api_base_url: str | None = None
    tool_llm_model_name: str | None = None
    tool_ollama_base_url: str | None = None
    tool_ollama_model: str | None = None
    # Intent Recognition AI config
    intent_llm_provider: str | None = None
    intent_llm_api_key: str | None = None
    intent_llm_api_base_url: str | None = None
    intent_llm_model_name: str | None = None
    # Emotion AI config (焕)
    emotion_llm_provider: str | None = None
    emotion_llm_api_key: str | None = None
    emotion_llm_api_base_url: str | None = None
    emotion_llm_model_name: str | None = None
    # Format AI config (遥)
    format_llm_provider: str | None = None
    format_llm_api_key: str | None = None
    format_llm_api_base_url: str | None = None
    format_llm_model_name: str | None = None
    show_tool_result: bool | None = None
    show_collaboration: bool | None = None
    token_monthly_budget: int | None = None


class SettingResponse(BaseModel):
    id: int
    user_id: int
    llm_provider: str | None = None
    llm_api_key: str | None = None
    llm_api_base_url: str | None = None
    llm_model_name: str | None = None
    ollama_base_url: str | None = None
    ollama_model: str | None = None
    # Tool Generation AI config
    tool_llm_provider: str | None = None
    tool_llm_api_key: str | None = None
    tool_llm_api_base_url: str | None = None
    tool_llm_model_name: str | None = None
    tool_ollama_base_url: str | None = None
    tool_ollama_model: str | None = None
    # Intent Recognition AI config
    intent_llm_provider: str | None = None
    intent_llm_api_key: str | None = None
    intent_llm_api_base_url: str | None = None
    intent_llm_model_name: str | None = None
    # Emotion AI config (焕)
    emotion_llm_provider: str | None = None
    emotion_llm_api_key: str | None = None
    emotion_llm_api_base_url: str | None = None
    emotion_llm_model_name: str | None = None
    # Format AI config (遥)
    format_llm_provider: str | None = None
    format_llm_api_key: str | None = None
    format_llm_api_base_url: str | None = None
    format_llm_model_name: str | None = None
    show_tool_result: bool | None = None
    show_collaboration: bool | None = None
    token_monthly_budget: int | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

    @field_validator("llm_api_key", "tool_llm_api_key", "intent_llm_api_key", "emotion_llm_api_key", "format_llm_api_key", mode="before")
    @classmethod
    def mask_api_key(cls, v: str | None) -> str | None:
        if v is None:
            return None
        if len(v) <= 8:
            return "****"
        return v[:4] + "***" + v[-4:]
